todo_app/app.py

```diff
@@ -28,26 +36,103 @@ def create_app():
         'DELETE':todoCards.delete_one
     }
 
+    #setup Login Manager
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        
+        state = str(uuid.uuid4())
+        session['state']=state
+        url = f"https://github.com/login/oauth/authorize?client_id={os.getenv('OAUTH_CLIENT_ID')}&state={state}"
+
+        #If login is disabled for testing we can go straight to home page
+        if os.getenv('LOGIN_DISABLED')=='True':
+            return redirect("/")
+        return redirect(url)
+
+    #Route when user not authenticated
+    @app.route('/login/callback')
+    def authenticate_user():
+        if (not 'state' in session) or (session['state'] != request.args['state']):
+                return render_template('unauthorised.html')
+        if "error" in request.args:
+            return render_template('unauthorised.html')
+        code = request.args['code']
+        url = "https://github.com/login/oauth/access_token"
+        headers = {
+            "Accept": "application/json"
+        }
+        params = {
+            "client_id": os.getenv('OAUTH_CLIENT_ID'),
+            "client_secret": os.getenv('OAUTH_CLIENT_SECRET'),
+            "code": code
+        }
+        resp = requests.request("POST",url,headers=headers,params=params)
+        if not resp.ok:
+            return render_template('unauthorised.html')
+      
+        access_token = resp.json()["access_token"]
+        headers['Authorization'] = "token " + access_token
+        url = "https://api.github.com/user"
+        user_response = requests.request("GET",url,headers=headers,params=params)
+      
+        if not user_response.ok:
+            return render_template('unauthorised.html')
+      
+        user_id = user_response.json()["id"]
+        user = User(user_id)
+        logged_in = flask_login.login_user(user)
+        
+        return redirect("/")    
+        
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
+        
+    def write_check(func):
+        @wraps(func)
+        def wrapper_write_check(*args, **kwargs):
+            
+            allowed = False
+            user = flask_login.current_user
+            
+            if os.getenv('LOGIN_DISABLED')=='True' or user.role == "edit":
+                allowed = True
+            if allowed:
+                return func(*args, **kwargs)
+            else:
+                return render_template('unauthorised.html')
```

Comment:
If you want to actually return a 401 instead of just saying 401 on the page, you can:

```python
return render_template('unauthorised.html'), 401
```